[ENH] Bep 005: Arterial Spin Labeling (#652)
bids-standard/bids-specification: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md

}
```

## Arterial Spin Labeling perfusion data

Template:

```Text
sub-<label>/[ses-<label>/]

>     perf/

Reviewer: What does perf stand for? Perfusion? Is asl the only modality using this? Maybe add a very brief line below describing and pointing where all these data folder labels are summarized

Author: Good point, @patsycle

Reviewer 2: I think these folders are the currently used datatype folders: https://github.com/bids-standard/bids-specification/tree/master/src/schema/datatypes

see definition of "datatype" here: https://bids-specification.readthedocs.io/en/latest/02-common-principles.html#definitions

Reviewer 3: @tsalo I saw perf is not yet in these files. I assume we (BEP005) should add some information there as well?

Reviewer: Thanks @sappelhoff @tsalo these links are super useful :)

Reviewer 4: Sorry for the delay- yes, updating the schema to incorporate ASL will have to happen before the PR is merged, but we are holding off on that part until the actual content and structure of the BEP have been approved. When the time comes, I can help.

Author: Do we need to help with the definition of "perf"? I would not restrict this to ASL, but for now this would only include ASL.

>     perf/

Author: Understandable confusion! We had the following arguments: 1) perfusion is a specific physiological quantity, 2) perfusion does not have to be a proxy of brain function, 3) (ASL) perfusion MRI has a very specific structure and metadata. Indeed, the community of ASL perfusion MRI is and has been more a combination of MRI physicists and radiologists, which differs from the usual BIDS community. I believe PET-BIDS will have similar arguments to not put e.g. FDG-PET under "func" but under a "pet" folder?

Whereas perfusion can indeed be used as an (arterial) proxy to measure brain function,  much like BOLD-fMRI (venous), perfusion differs as this provides a physiological parameter that can be influenced by more processes than only functional, and hence used for more applications than only as a proxy of brain metabolism/function. ASL perfusion MRI quantifies how much mL blood flows through 100 grams of brain tissue per minute, which is not only dependent on the demand of blood flow - which could be seen as a functional proxy - but also very much on the ability of blood supply. Indeed, ASL's most clinical applications are cerebrovascular - e.g. how much blood can still be supplied to tissue in the case of proximal arterial obstructions - and oncology - e.g. angiogenesis (not necessarily metabolism) could indicate tumor malignancy.  The "perf" directory will contain data that measure perfusion as a physiological quantity, whereas all techniques currently in the "func" directory, are arbitrary units of which it's derivatives can estimate brain function. 

Are these arguments convincing? @patsycle will be writing a paper containing these and other motivations for the design decisions we made.

>     perf/

Reviewer 5: Thanks for the clarifications.

>     perf/

Reviewer: Thanks! Maybe include briefly some of these ideas in the draft?

>     perf/

Reviewer 3: @guiomar Will do! Added it on my to do list. Thanks!

>     perf/

Author: @guiomar: Did you mean adding this to the paper or to the specification? We can add it very shortly to the specification but I thought this should not have much explanation of design choices?

       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_asl.nii[.gz]
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_asl.json
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_aslcontext.tsv
       [sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_m0scan.nii[.gz]]
       [sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_m0scan.json]

>        [sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_labeling.jpg]

Author: I see. @patsycle let's make sure that we explicitly state in the text body above this template that these files are optional, and remove the brackets in the template.

```

> The complete ASL timeseries should be stored as a 4D NIfTI file in the original acquisition order, accompanied by two ancillary 

Reviewer: ```suggestion
**`*_asl.nii[.gz]`:** The complete ASL timeseries should be stored as a 4D NIfTI file in the original acquisition order, accompanied by two ancillary 
```
I like when all the different files are highlighted because I can find them easily. But it's a personal preference, ignore it if you don't like

> The complete ASL timeseries should be stored as a 4D NIfTI file in the original acquisition order, accompanied by two ancillary 

Author: @patsycle my spelling control (Grammarly) also tells me we should use `time series` or `time-series` instead of `timeseries`. Let's choose the one that is most correct (in US spelling - if I'm correct this is the English spelling that BIDS uses?) and use this one throughout.

> The complete ASL timeseries should be stored as a 4D NIfTI file in the original acquisition order, accompanied by two ancillary 

Reviewer 3: done

> files: `*_asl.json` and `*_aslcontext.tsv`. The `*_asl.nii.gz` and `*_m0.nii.gz` should contain appropriately scaled data and no scaling factors 

Reviewer: ```suggestion
files: `*_asl.json` and `*_aslcontext.tsv`. The `*_asl.nii.gz` and `*_m0scan.nii.gz` should contain appropriately scaled data and no scaling factors 
```
Do you refer to the m0scan, or is it another file?

Author: yes you're right!

should be stored in the ancillary files or in the NIfTI header. 

The `m0scan` can either be stored inside the 4D ASL time-series NIfTI file or as a separate NIfTI file, depending on whether it 

> was acquired within the ASL time-series or as a separate scan. These and other M0 options are specified in the required M0 field. 

Reviewer: ```suggestion
was acquired within the ASL time-series or as a separate scan. These and other M0 options are specified in the required `M0` field of the `*_asl.json` file. 
```
The M0 field is in the asl.json file, right?

Author: Yes

> It can also be stored under `fmap/sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.nii[.gz]`, 

Author: We should indeed be consistent. In ASL quantification (and jargon), the M0 is the reference magnetization (in the same a.u. as the `control` and `label` acquisition) that a voxel with 100% blood protons should have. There are several ways to estimate this, which all have pros and cons, one of which is to use an `m0scan`. This `m0scan` is not a true `M0` but a practical proxy, which is why has been previously defined as the recommended choice for ASL quantification. We wanted to keep this distinction between the `M0` and `m0scan` explicit in ASL BIDS, to help the users. The `M0` JSON field now states which type of M0 information was acquired. Perhaps we can make this clearer by e.g. renaming this JSON field to `M0Type`. @patsycle let's discuss this internally.

> It can also be stored under `fmap/sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.nii[.gz]`, 

Reviewer 3: Indeed, as Henk mentions, the m0scan is when there is a scan which can be used for calibration. Alternatively, you can use a value or the mean control for calibration, however, this is always referred to as M0 within the field.. Thus, I would not change the naming of the scan, but we might rethink the M0 field, as already suggested in another comment. @effigies you are right we used m0.nii, however this appears to be a typo. When talking about the scan or scan volume, it should always be m0scan. Thanks for noticing this.

when the pepolar approach is used.

> 

Reviewer: ```suggestion

### `*_aslcontext.tsv`
```
Add heading for this data file

> The `*_aslcontext.tsv` lists the volume_type for each volume in a single column, in the `*_asl.nii[.gz]`, corresponding to the volume_types specified in the table below, based 

Reviewer: ```suggestion
## `*_aslcontext.tsv`

The `*_aslcontext.tsv` lists the volume_type for each volume in a single column, in the `*_asl.nii[.gz]`, corresponding to the volume_types specified in the table below, based 
```

| label           | The label image is acquired in the exact same way as the control image, except that the blood magnetization flowing into the imaging region has been inverted.                         |
| m0scan          | The M0 image is a calibration image, used to estimate the equilibruim magnetization of blood.                                                                                          |
| deltam          | The deltaM image is a perfusion-weighted image, obtained by the subtraction of `control` - `label`.                                                                                    |
| cbf             | The cerebral blood flow (CBF) image is produced by dividing the deltaM by the M0, quantified into `mL/100g/min` (See also [doi:10.1002/mrm.25197](https://doi.org/10.1002/mrm.25197)). |

If the raw `control` and `label` images are not available, their derivative `deltam` should be stored as raw data instead. If the 
`deltam` is not available, `cbf` should be stored as raw data instead. When `cbf` is stored as raw data, its units need to be 

> specified in the `*_asl.json` as well.

Reviewer 4: As mentioned in https://github.com/bids-standard/bids-specification/pull/652/files#r516908561, this description can be confusing. I think it's important to explicitly link `deltam` and `cbf` as volume types (rather than modalities or suffixes) to the `asl.nii.gz` file. A good way to do that would be to replace "raw data" (the whole BIDS dataset is essentially "raw data") with "`_asl.nii.gz`".

E.g., "their derivative `deltam` should be stored in the `_asl.nii.gz` file instead."

Author: Yes this makes sense, thanks!

> `*_aslcontext.tsv` example:

Reviewer 4: It would be helpful to see examples for `deltam` and `cbf` cases. It might even be worth it to organize this section similarly to [the fmap section](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#fieldmap-data) (i.e., as a limited set of mutually exclusive cases, each with its own description and example).

Author: Yes, this makes sense, @patsycle we can check how this would look?

Reviewer 3: @tsalo I commited three cases. Feel free to take a look and give some feedback!

Reviewer 4: The cases are very helpful, thank you. I have a couple of notes that I will follow up with at the specific lines.

| **volume_type** |
|-----------------|
| m0scan          |
| control         |
| label           |
| control         |
| label           |
| control         |

> | label           |

Author: You may be right, but we thought that a Table would be clearer, as this is an example of how the _aslcontext.tsv file would look like. We had long discussions over whether we should concatenate horizontally or vertically, and vertically (a column of volume_types) seemed easier to read for users.

> ### Metadata fields applicable to both (P)CASL and PASL 

Reviewer: ```suggestion
### `*_asl.json` file

#### Metadata fields applicable to both (P)CASL and PASL 
```

Does this relates to the asl.json file?

Author: Yes, but this is also the case for the (P)CASL- and PASL-specific metadata fields. We wanted to clarify which fields are applicable to both LabelingTypes or only to a specific LabelingType, hence why your suggested replacement could be confusing.

> | LabelingType                      | REQUIRED                                                                                                                        | [string][]                              | `CASL` , `PCASL` , `PASL`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Author: Good point that the word `labeling` can be ambiguous. In ASL jargon, `ASLType` could be seen broader than only the type of labeling, e.g. Single PLD PCASL and multi PLD PCASL are two different types of ASL. Perhaps `LabelingPulseType` is the least potentially ambiguous wording. @patsycle let's discuss this internally in the ASL-BIDS group.

> | LabelingType                      | REQUIRED                                                                                                                        | [string][]                              | `CASL` , `PCASL` , `PASL`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Reviewer 3: @effigies We decided to change this field to ArterialSpinLabelingType to be consistent with the ASL lexicon.

> | PostLabelingDelay                 | REQUIRED                                                                                                                        | [number][] or [array][] of [numbers][]  | This is the time, in seconds, after the the end of the labeling (for `(P)CASL`) or middle of the labeling pulse (for `PASL`) until the middle of the excitation pulse applied to the imaging slab (for 3D acquisition) or first slice (for 2D acquisition).  Can be a number (for a single-PLD time series) or an array of numbers (for multi-PLD and Look-Locker). In the latter case, the array of numbers contains the PLD of each volume, namely each `control` and `label`, in the acquisition order. Any image within the time-series without a PLD, for example an `m0scan`, is indicated by a zero. Based on DICOM Tags 0018,9079 `Inversion Times` and 0018,0082 `InversionTime`.                                                                                                                                                                                                                                                                                                                                                                                                          |

Reviewer: Define PLD acronym, is it PostLabelingDelay? I think it's been never defined.
Also a summary figure like in the PET extension may help to visualize some parameters

Author: Yes thanks!

> | M0                                | REQUIRED                                                                                                                        | [number][] or [string][] or [boolean][] | Describes the available M0 information or the location of an `m0scan` file, as either: *Numerical*: a single numerical whole-brain M0 value, for example obtained from CSF in another sequence; *SeparateFile*: `*_m0scan.nii[.gz]`; *Within ASL time-series*: *Boolean: `true`* referring to an `m0scan` within the time-series (as defined within the `*_aslcontext.tsv`); *`false`* referring to the use of the (average) `control` image for M0 calibration.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer: ```suggestion
| M0                                | REQUIRED                                                                                                                        | [number][] or [string][] or [boolean][] | Describes the available M0 information or the location of an `m0scan` file, as either: 
- *Numerical*: a single numerical whole-brain M0 value, for example obtained from CSF in another sequence;
- *SeparateFile*: `*_m0scan.nii[.gz]`; 
- *Within ASL time-series*: 
- *Boolean: `true`* referring to an `m0scan` within the time-series (as defined within the `*_aslcontext.tsv`);
- *Boolean: `false`* referring to the use of the (average) `control` image for M0 calibration.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
```

Reviewer 5: From separate discussion of `M0.nii.gz`, this might change from `M0`. What I've suggested above would be `M0Estimate`. You could also add an explicit `M0Type` that could be a string such as "Scan", "Volume", "Estimate", corresponding to current "SeparateFile`, `true`, `<number>` to indicate which place to look, if the implicit logic is uncomfortable.

Overall point is that I think two variables are better than one highly overloaded one.

Reviewer 3: @effigies  Thanks for the feedback and suggestions. Indeed, I tried the bullet points, but that doesn't work with table. 
I understand that the current options for the M0 might be a bit much. However, I feel that the strategy 'M0 optional' you proposed might lead to datasets for which M0 is not defined correctly. Additionally, a seperately acquired scan should always be preferred. However, splitting up the fields might be a better option. I will definitely put this on the agenda of the meeting this week. Let's see if we can find another solution for this, based on your suggestions.

Author: I like the solution `M0Type`: "Scan", "Volume", "Estimate" (we don't need to specify the "fallback to average control image"). And then we have indeed a separate optional numeric field `M0Estimate` that is required when M0Type equals 'Estimate'. @patsycle let's indeed discuss this internally

> | EchoTime                          | REQUIRED                                                                                                                        | [number][] or [array][] of [numbers][]  | The echo time (TE) for the acquisition, specified in seconds. Specify either one number for the total time-series, or provide an array of numbers with different values of each volume. Corresponds to DICOM Tag 0018, 0081 `Echo Time` (please note that the DICOM term is in milliseconds not seconds).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer 4: This seems to conflict with existing approaches to multi-echo data (i.e., using the echo entity and storing the data in separate files). Do the separate echoes need to be in the same file?

> | EchoTime                          | REQUIRED                                                                                                                        | [number][] or [array][] of [numbers][]  | The echo time (TE) for the acquisition, specified in seconds. Specify either one number for the total time-series, or provide an array of numbers with different values of each volume. Corresponds to DICOM Tag 0018, 0081 `Echo Time` (please note that the DICOM term is in milliseconds not seconds).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer: where is "(TE)" definition used?

> | SliceTiming                       | RECOMMENDED, but REQUIRED in case of a 2D acquisition, or when VolumeTiming is defined                                          | [array][] of [numbers][]                | The time at which each slice was acquired within each volume (frame) of the acquisition with respect to the beginning of volume acquisition, namely after the PostLabelingDelay. In the case of 3D acquisition, a zero is entered. The length of this array of numbers should be equal to the number of slices for 2D acquisition, with the first time being zero, effectively equal to the PostLabelingDelay. Slice timing is not the same as slice order. Rather, it is a list of times containing the time (in seconds) of each slice acquisition in relation to the beginning of volume acquisition. The list goes through the slices along the slice axis in the slice encoding dimension. Note that to ensure the proper interpretation of the `SliceTiming` field, it is important to check if the OPTIONAL `SliceEncodingDirection` exists. In particular, if `SliceEncodingDirection` is negative, the entries in `SliceTiming` are defined in reverse order with respect to the slice axis, such that the final entry in the `SliceTiming` list is the time of acquisition of slice zero. |

Reviewer: ```suggestion
| SliceTiming                       | RECOMMENDED, but REQUIRED in case of a 2D acquisition, or when `VolumeTiming` is defined                                          | [array][] of [numbers][]                | The time at which each slice was acquired within each volume (frame) of the acquisition with respect to the beginning of volume acquisition, namely after the `PostLabelingDelay`. In the case of 3D acquisition, a zero is entered. The length of this array of numbers should be equal to the number of slices for 2D acquisition, with the first time being zero, effectively equal to the `PostLabelingDelay`. Slice timing is not the same as slice order. Rather, it is a list of times containing the time (in seconds) of each slice acquisition in relation to the beginning of volume acquisition. The list goes through the slices along the slice axis in the slice encoding dimension. Note that to ensure the proper interpretation of the `SliceTiming` field, it is important to check if the OPTIONAL `SliceEncodingDirection` exists. In particular, if `SliceEncodingDirection` is negative, the entries in `SliceTiming` are defined in reverse order with respect to the slice axis, such that the final entry in the `SliceTiming` list is the time of acquisition of slice zero. |
```

> | AcquisitionVoxelSize              | RECOMMENDED                                                                                                                     | [array][] of [numbers][]                | An array of numbers with a length of 3, in millimeters. This parameter denotes the original acquisition voxel size, excluding any inter-slice gaps and before any interpolation or resampling within reconstruction or image processing. Any point spread function effects, for example due to T2-blurring, that would decrease the effective resolution are not considered here.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |

Reviewer 4: I think this could fall under "Common metadata fields" for MRI in general.

Reviewer 6: I might be wrong but I think that @tsalo was also referring to the fact that several items in this table that already exist in the "common metadata" or another table.

https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#common-metadata-fields

https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#sequence-specifics

https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#timing-parameters

For example (I have not checked them all):

- `MagneticFieldStrength` is in the common-metadata-fields table.
- `PulseSequenceType` already exists in sequence-specifics table but the description is different.
- `EchoTime` is in the timing parameters table.
- I suspect `VolumeTiming`, `SliceTiming`, `RepetitionTime` are in the same boat.

I think that if keyname already exists somewhere else AND the description is the same, then it might not be necessary to duplicate them here. If only the requirement level change maybe we can have a lighter version of the table.

I am not sure how I feel about keeping the same keyname if the description for ASL is different than in the field that already exists.

Reviewer 4: > I think that if keyname already exists somewhere else AND the description is the same, then it might not be necessary to duplicate them here. If only the requirement level change maybe we can have a lighter version of the table.

Once we have rendering with the schema fully set up, tables with truly duplicated definitions should be easy enough to build. I think it's just key that we make sure that the definitions are completely the same.

> I am not sure how I feel about keeping the same keyname if the description for ASL is different than in the field that already exists.

I agree. I'm not sure if there's a written rule about this, but my opinion is that descriptions or formats of metadata fields should not vary  depending on the section of the specification, although requirement levels can.

Reviewer 3: Hi @tsalo and @Remi-Gau. This is an issue with which were struglin with for a while, and contacted BIDS maintainers for several times. The definitions are always the same, but there are some additional specifics for ASL. For example, MagneticFieldStrength becomes required (which is needed for the quantifcation of ASL data), PulseSequenceType has to include 2D or 3D, in order to allow conditional check for SliceTiming, etc. I can easily create a document comparing the differences between the definition for 'common' and for 'asl' for each duplicated field, if that helps? Maybe we should go over them one by one, together with some of the maintainers?

Reviewer 6: @patsycle Sorry I just joined the maintainers recently and I did not realize there had been some previous discussions about this so I jumped in a bit too abruptly. :pray: 

I actually do think that have this "difference" document between ASL and common would be useful: it really makes me realize I read this section way too quickly (or that I am just bad at playing "spot the differences" :wink: )

Reviewer 3: @Remi-Gau Haha, no problem. This is, in my opinion, one of the most difficult parts of the extension, and probably a 'main bids topic' which should be cleared out in view of other extensions as well. So I am glad that this question has been raised! I will create this document tomorrow or on Thursday and add the link here. Maybe some kind of 'rule of thumb' can be created this way.

Author: Yes @patsycle let's redefine these main parameters in this PR such that it easy to see the difference. I don't think we need a separate document.

> | FlipAngle                         | RECOMMENDED, REQUIRED for ASL-sequences that use a Look-Locker readout and/or multiple flip angles (for example (turbo-)QUASAR) | [number][] or [array][] of [numbers][]  | Flip angle for the acquisition, specified in degrees, either as a number or an array of numbers, with the length corresponding to the number of volumes. Corresponds to: DICOM Tag 0018, 1314 `Flip Angle`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |

Reviewer 4: This seems to conflict with other approaches to data with multiple flip angles (e.g., the `VFA` suffix proposed in BEP001), where a new `flip` entity is added and the data are separated into different files. Do `(turbo-)QUASAR` data _need_ to be in one file?

Reviewer 3: Hi @tsalo . Indeed, the approach used for BEP005 is different from those used by BEP001. We had a discussion with the BEP001 about this before submitting the PR, and we agreed that for this case, not one size fits all. 

For their case, most volumes are already stored as separate nii files by the scanner, whereas in our case, all volumes are stored in one nii file (timeseries). Both groups agreed that the best approach is to keep the datafiles as they are acquired, so no merging, and also no splitting up. Additionally, for ASL, it is very important not to split up the file, for processing purposes (e.g. motion correction), as the data files are already very small. This applies also to QUASAR data, where, like in other multi-PLD ASL files, all data are kept in a single file as acquired.

Reviewer 4: It does sound like this is key for ASL, but changing the format and definition of the field in the ASL section means changing it for the specification as a whole. We discussed this issue in today's BIDS maintainers call, and we don't want to change the format for a common metadata field without discussing it more in the context of other parts of the specification. 

As a hack, we're proposing that variable EchoTimes and FlipAngles _not_ be set in the json file, and instead be set in the aslcontext.tsv file as additional columns. Tagging @bids-standard/maintainers, who were involved in the proposal.

Reviewer 3: @tsalo I will propose this in our meeting later this week.

Author: That's a nice suggestion. To clarify, it is not about the storage by the scanner, but it is about whether scans fall in the same or a different sequence as acquired by the scanner, and to keep acquisition order intact. I'm not sure if for QUASAR this is an absolute requirement, we chose this for consistency.

> | VascularCrushingVenc              | OPTIONAL, RECOMMENDED if `VascularCrushing` is `true`                                                                           | [number][] or [array][] of [numbers][]  | The crusher gradient strength, in centimeters per second. Specify either one number for the total time-series, or provide an array of numbers, for example when using QUASAR, using the value zero to identify volumes for which `VascularCrushing` was turned off. Corresponds to DICOM Tag 0018,925A `ASL Crusher Flow Limit`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |

Author: Yes you're right, this should be VascularCrushingVENC

> ### (P)CASL-specific metadata fields

Reviewer: ```suggestion
#### (P)CASL-specific metadata fields
```

| LabelingPulseInterval        | RECOMMENDED           | [number][]                             | Delay between the peaks of the individual labeling pulses, in seconds.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |

> ### PASL-specific metadata fields

Reviewer: ```suggestion
#### PASL-specific metadata fields
```

| BolusCutOffDelayTime  | OPTIONAL, REQUIRED if `BolusCutOffFlag` is `true` | [number][] or [array][] of [numbers][] | Duration between the end of the labeling and the start of the bolus cut-off saturation pulse(s), in seconds. This can be a number or array of numbers, of which the values must be non-negative and monotonically increasing, depending on the number of bolus cut-off saturation pulses. For Q2TIPS, only the values for the first and last bolus cut-off saturation pulses are provided. Based on DICOM Tag 0018,925F `ASL Bolus Cut-off Delay Time`. |
| BolusCutOffTechnique  | OPTIONAL, REQUIRED if `BolusCutOffFlag` is `true` | [string][]                             | Name of the technique used, for example Q2TIPS, QUIPSS, QUIPSSII. Corresponds to DICOM Tag 0018,925E `ASL Bolus Cut-off Technique`.                                                                                                                                                                                                                                                                                                                     |

> ### m0scan-specific metadata fields

Reviewer: ```suggestion
### `m0scan` metadata fields
```

|-------------------------------|---------------------------------|-------------------------------|-------------------------------|
| PulseSequenceType             | 2D acquisition / 3D acquisition | SliceTiming                   | \[X\] / \[\]                  |
| LookLocker                    | true                            | FlipAngle                     | \[X\]                         |
| M0                            | true / false                    | `*_aslcontext.tsv`            | Contains `m0scan` / `control` |
| LabelingType                  | PASL                            | BolusCutOffFlag               | \[X\]                         |
| BolusCutOffFlag               | true / false                    | BolusCutOffDelayTime          | \[X\] / \[\]                  |
| BolusCutOffFlag               | true / false                    | BolusCutOffTechnique          | \[X\] / \[\]                  |
| `*_aslcontext.tsv`            | cbf                             | Units                         | \[X\]                         |
| RepetitionTime / VolumeTiming | \[X\] / \[\]                    | VolumeTiming / RepetitionTime | \[\] / \[X\]                  |

> | VolumeTiming                  | \[X\]                           | AcquisitionDuration           | \[X\]                         |

Author: A flow-chart could be clearer, but I believe BIDS already had similar Tables in other sections, which is why we formatted this way.

> | VolumeTiming                  | \[X\]                           | AcquisitionDuration           | \[X\]                         |

Reviewer 5: Yeah, there is a entity/suffix compatibility matrix, which TBH did not translate well to Markdown.

If you're up for putting together flow charts, that would be worth considering. The PET extension also has figures, so there's precedent.

> | VolumeTiming                  | \[X\]                           | AcquisitionDuration           | \[X\]                         |

Reviewer 3: I don't see how we could get this into a flowchart, nor figure ;) This is more a summary for what is already defined within the BIDS fields themselves. Additionally, this is also implemented in the validator, so if done wrong, you get an error anyway (with additional information).

> | VolumeTiming                  | \[X\]                           | AcquisitionDuration           | \[X\]                         |

Author: We can try to convert this dependency table to a flowchart, let's see how the PET guys did this.

> | VolumeTiming                  | \[X\]                           | AcquisitionDuration           | \[X\]                         |

Reviewer 3: We decided to keep this dependency list, as there is already another in the specification. Additionally, we are creating a flowchart which will be included in the appendix.

**Legend**

-   \[ X ] --> has to be filled

> -   \[   \] --> has to be left empty

Reviewer 6: I suspect that this was taken from the table about slice timing: https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#timing-parameters_1

We should also probably keep that other table consistent with this one then.

}
```

## Arterial Spin Labeling perfusion data

Template:

```Text
sub-<label>/[ses-<label>/]
    perf/
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_asl.nii[.gz]
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_asl.json
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_aslcontext.tsv
       [sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_m0scan.nii[.gz]]
       [sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_m0scan.json]

>        [sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_labeling.jpg]

Reviewer 4: I think the `deltam` and `cbf` suffixes should be represented in this template as well.

Reviewer 3: We provided a table with volume types, which are often scanned sequentially in ASL. These can include any combination of raw volumes (control, label, and/or m0scan) and/or derivative volumes (deltam and/or cbf). To keep the original acquisition order intact (e.g. for subsequent motion correction and other temporally-specific artifacts) we opted to keep a single 4D NIfTI that contains any combination of volume types, which is then specified in the _asl_context.tsv sidecar. While deltam and cbf may indeed become separate derivative ASL NIfTIs, it may become confusing to also allow this as a suffix to the raw ASL NifTI. The reason to add the option to allow these derivatives deltam and cbf as raw volumes is that some vendor ASL sequences do not provide the original control-label raw data. In those cases, the derivatives deltam and cbf will (also) be seen as ‘raw’ data, as it is the only data available. Is this currently described with sufficient clarity?

Reviewer 4: I think the logic for storing makes sense, but since they will still be valid files with their own rules regarding sets of entities and metadata, they should be reflected in a "Template". That template could possibly go below, where cbf and deltam are discussed in more detail, but I think it's just clearer to have them here with everything else.

Reviewer 3: The files do not exist yet in this extension. Only in case there is no true raw data (control, label), the deltam or cbf can be used. However, even in this case, the file should always be called: *_asl.json. The distinction between control and label versus deltam or cbf in this case, is done in the aslcontext.tsv. The use of *_deltam or *_cbf should be supported in a (future) ASL-BIDS derivatives extension, so to avoid confusion, I think it is more logic not to mention it in the templates. Also, the validator will not yet agree with those files.

Reviewer 4: Just to make sure I understand this, `cbf` and `deltam` would be volumes in the `_asl.nii.gz` file?

Reviewer 3: Indeed. In most cases, the *_asl.nii will be one file with for example control-label-control-label-m0scan volumes, which you logically define in the aslcontext. Sometimes, additional files can be retrieved, such as a cbf image (calculated CBF so derivative), which should be stored in derivatives but is not yet included in this extension. However, some sequences (for example for GE) don't provide these raw data volumes, but already do the quantification online. So when getting your asl from the scanner, this will be for example one file with only a cbf volume. In this case (lack of true raw data), you can store your cbf image as *_asl.nii, with mentioning cbf as volume_type in the aslcontext. 

I hope this clarifies it?

Reviewer 4: Yes, it does, thank you! I was misinterpreting your original explanation. In that case, you're totally right not to include nonexistent suffixes in the template here. I think the description of those volume types below could use some clarification, though, so others don't end up confused like me. I will bring that up at the relevant lines.

specified in the table below, based on DICOM Tag (0018,9257) `ASL Context`. Note that the volume_types `control` and 
`label` within BIDS do not discriminate between types of labeling, but only serve to specify the magnetization state of the 
blood and thus the ASL subtraction order.

| volume_type                   | Definition    |
|---------------------------------- | -----------------------------------------------------|
| control                  |  The control image is acquired in the exact same way as the label image, except that the magnetization of the blood flowing into the imaging region has not been inverted.  |
| label                  |The label image is acquired in the exact same way as the control image, except that the blood magnetization flowing into the imaging region has been inverted. |
| m0scan                  | The M0 image is a calibration image, used to estimate the equilibruim magnetization of blood.        |
| deltam                  | The deltaM image is a perfusion-weighted image, obtained by the subtraction of `control` - `label`.  |

> | cbf                  | The cerebral blood flow (CBF) image is produced by dividing the deltaM by the M0, quantified into `mL/100g/min` (See also <https://pubmed.ncbi.nlm.nih.gov/24715426/>).     |

Reviewer 2: let's use DOIs wherever possible. My suggestion will render like this: [doi:10.1002/mrm.25197](https://doi.org/10.1002/mrm.25197).

Note that after accepting this suggestion, table pipes will have to be adjusted again.

```suggestion
| cbf                  | The cerebral blood flow (CBF) image is produced by dividing the deltaM by the M0, quantified into `mL/100g/min` (See also [doi:10.1002/mrm.25197](https://doi.org/10.1002/mrm.25197)).     |
```

|  LabelingLocationDescription     | RECOMMENDED   | [string][]    | Description of the location of the labeling plane (`(P)CASL`) or the labeling slab (`PASL`) that cannot be captured by fields ‘LabelingOrientation’ or ‘LabelingDistance’. May include a link to an anonymized screenshot of the planning of the labeling slab/plane with respect to the imaging slab or slices `sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_labeling.jpg`. Based on DICOM macro C.8.13.5.14.  |
| LookLocker  | OPTIONAL | [boolean][]  | Boolean indicating if a Look-Locker readout is used.| 
|  LabelingEfficiency     | OPTIONAL   | [number][]  | Labeling efficiency, specified as a number between zero and one. |

### (P)CASL-specific metadata fields
These fields can only be used when `LabelingType` is `CASL` or `PCASL`.
| **Key name**                  | **Requirement level** | **Data type** | **Description**                                                   |

> |-------------------------------|-----------------------|---------------|------------------------------------------------------------------------------------------|

Reviewer 4: This table appears malformed. It doesn't render in the RTD CI.

Reviewer 2: fyi @patsycle @HenkMutsaerts "RTD CI" means "Readthedocs continuous integration service" and Taylor refers to the following GitHub check listed at the end of this PR (see screenshot):

![image](https://user-images.githubusercontent.com/9084751/97070524-484ca080-15d9-11eb-918f-b8567919805f.png)

by clicking on it, you can see the draft rendering of the PR, like [here](https://bids-specification--652.org.readthedocs.build/en/652/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#pcasl-specific-metadata-fields). --> There you can see that some table formatting seems to have gone awry.

Note that after each commit, you have to click again on the check at the end of the PR to get the up to date link

Author: Thanks, yes, that shouldn't be a problem to adjust.

Reviewer 3: Table is fixed, but I still get some errors on a link. However, I cannot seem to find what the issue is. Any suggestions?

Reviewer 2: @patsycle some of your links are not formatted in markdown style, see this example:

```markdown
[this-link]
(http://www.wont-work.net)

[this-link](http://www.wont-work
-either.net)

[but-this-link](http://www.will-work.success)
```

Reviewer 2: That is, a link in markdown format must be on a single line

> |  LabelingDuration                 | REQUIRED            | [number][] or [array][] of [numbers][]         | Total duration of the labeling pulse train, in seconds, corresponding to the temporal width of the labeling bolus for `(P)CASL`. Specify either one number for total time-series or provide an array of numbers of different values for each volume in case of sequential acquisitions with varying labeling duration. In the case that an array of numbers is provided, its length should be equal to the number of volumes specified in  `sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_aslcontext.tsv`. An `m0scan` acquired within the ASL time-series, gets a value of zero. Corresponds to DICOM Tag 0018,9258 `ASL Pulse Train Duration`. |

Reviewer 4: ```suggestion
|  LabelingDuration                 | REQUIRED            | [number][] or [array][] of [numbers][]         | Total duration of the labeling pulse train, in seconds, corresponding to the temporal width of the labeling bolus for `(P)CASL`. Specify either one number for total time-series or provide an array of numbers of different values for each volume in case of sequential acquisitions with varying labeling duration. In the case that an array of numbers is provided, its length should be equal to the number of volumes specified in the associated `aslcontext.tsv`. An `m0scan` acquired within the ASL time-series, gets a value of zero. Corresponds to DICOM Tag 0018,9258 `ASL Pulse Train Duration`. |
```

This will be easier to read, unless there is a reason I'm missing to include the full name.

Reviewer 4: It would be great to apply this to all of the instances of full filenames in these tables as well.

Reviewer 3: done

|  LabelingPulseMaximumGradient                  | RECOMMENDED             | [number][]         | The maximum amplitude of the gradient switched on during the application of the labeling RF pulse(s), in milliteslas per meter. |
|  LabelingPulseAverageB1                   | RECOMMENDED             | [number][]         | The average B1-field strength of the RF labeling pulses, in microteslas. As an alternative, `LabelingPulseFlipAngle` can be provided.  |
|  LabelingPulseDuration                    | RECOMMENDED             | [number][]         | Duration of the individual labeling pulses, in seconds. |
| LabelingPulseFlipAngle | RECOMMENDED  | [number][]         | The flip angle of a single labeling pulse, in degrees, which can be given as an alternative to `LabelingPulseAverageB1`. |
|  LabelingPulseInterval                     | RECOMMENDED             | [number][]         | Delay between the peaks of the individual labeling pulses, in seconds. |

### PASL-specific metadata fields
These fields can only be used when `LabelingType` is `PASL`.

> | **Key name**                  | **Requirement level** | **Data type** | **Description**                                                   |

Reviewer 4: This table appears malformed. It doesn't render in the RTD CI.

Author: Thanks, we will adapt.

Reviewer 3: done

|-------------------------------|---------------------------------|-------------------------------|-------------------------------|
| PulseSequenceType             | 2D acquisition / 3D acquisition | SliceTiming                   | [X] / []                      |
| LookLocker                    | true                            | FlipAngle                     | [X]                           |
| M0                            | true / false                    | `*_aslcontext.tsv`            | Contains `m0scan` / `control` |
| LabelingType                  | PASL                            | BolusCutOffFlag               | [X]                           |
| BolusCutOffFlag               | true / false                    | BolusCutOffDelayTime          | [X] / []                      |
| BolusCutOffFlag               | true / false                    | BolusCutOffTechnique          | [X] / []                      |
| `*_aslcontext.tsv`            | cbf                             | Units                         | [X]                           |
| RepetitionTime / VolumeTiming | [X] / []                        | VolumeTiming / RepetitionTime | [] / [X]                      |

> | VolumeTiming                  | [X]                             | AcquisitionDuration           | [X]                           |

Reviewer 2: this table is problematic in terms of markdown syntax. the square brackets need to either be escaped using backslashes like `\[ X \]`, or they need to be wrapped in code syntax using backticks:

```
`[X]`
```

Reviewer 3: Ah it might indeed be this one. I'll try with your tips. Thanks

Reviewer 2: yes, you last commit 9f752fc fixed the Travis linter! :-)

Reviewer 3: YEs thanks! I was confused as the line reference was confusing. Anyway, learned a lot today!

Reviewer 3: Stefan, still problems with 2 tables in the readthedocs. Any ideas?

Reviewer 2: > Stefan, still problems with 2 tables in the readthedocs. Any ideas?

Can you provide some links and/or screenshots, please? :)

Reviewer 3: https://bids-specification--652.org.readthedocs.build/en/652/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#pcasl-specific-metadata-fields

![image](https://user-images.githubusercontent.com/41481345/97172924-0ade4380-1790-11eb-99e5-64c23977b09c.png)
![image](https://user-images.githubusercontent.com/41481345/97172954-1af62300-1790-11eb-8046-0b4024475a31.png)

Reviewer 2: @patsycle I think you are looking at the wrong "build artifact". The draft gets rebuilt after each commit that you do. This is the most recent one: https://bids-specification--652.org.readthedocs.build/en/652/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#pcasl-specific-metadata-fields

To get the most recent artifact, scroll to the bottom of the pull request user interface, and click on "Details" for the "docs/readthedocs.org:bids-specification" check, see this screenshot:

![image](https://user-images.githubusercontent.com/9084751/97175587-087de880-1794-11eb-87da-7f9081b30f57.png)

Reviewer 3: It seems I was looking at the right one, but the link was still in cache. Using a different browser worked. Thanks for all you help!

}
```

## Arterial Spin Labeling perfusion data

Template:

```Text
sub-<label>/[ses-<label>/]
    perf/
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_asl.nii[.gz]
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_asl.json
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_aslcontext.tsv
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_m0scan.nii[.gz]
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_dir-<label>][_run-<index>]_m0scan.json
       sub-<label>[_ses-<label>][_acq-<label>][_rec-<label>][_run-<index>]_labeling.jpg
```

The complete ASL time series should be stored as a 4D NIfTI file in the original acquisition order, accompanied by two ancillary 
files: `*_asl.json` and `*_aslcontext.tsv`. The `*_asl.nii.gz` and `*_m0scan.nii.gz` should contain appropriately scaled data and no scaling factors 

> should be stored in the ancillary files or in the NIfTI header. 

Reviewer 7: Sorry if I missed if and where this was discussed: what is the rationale for banning scaling factors in NIfTI headers? Does it mean `scl_slope` can only be `1.0`? From my naive viewpoint, it seems to be prescriptive of a level that should be abstracted away.

Reviewer 3: Hi @gllmflndn, this was indeed already discussed before, so I will copy @HenkMutsaerts 's response here. Feel free to ask your remaining questions:

ASL BIDS certainly takes the scale factors in account. We considered two options, of which we chose the second, as it has less potential for confusion.

1. keep image values as is, retain the scale factors, such that scaling needs to be dealt with in post processing
2. use the scale factors to scale the image values, removing the need for scaling in post processing.

We chose that the ASL BIDS standard should not have scale factors, which puts the responsability for correct scaling to the DICOM2BIDS conversion stage, when any incorrect scaling can still be dealt with, leaving no confusion as to whether scaling has or has not been applied in the BIDS data.

Hope this explains it!

Reviewer 7: Thank you @patsycle for letting me know and sorry for not having seen this response before.

I am still a bit confused though - if this restriction remains, I think it should be clarified what `"appropriately scaled data and no scaling factors should be stored in the ancillary files or in the NIfTI header"` exactly means. If this means `scl_slope` and `scl_inter`  in the NIfTI header have to be `1` and `0` respectively then it should be explicit. A DICOM to NIfTI converter would rarely change the data type while it means here that `datatype` will almost always be `float` or `double`.
Why is there a mention of `ancillary files` regarding scaling as there doesn't seem to be any scale-related parameters anyway?

And apologies if I misunderstand something or if it's too late to have this conversation but I don't see what is specific with ASL data that would require a restriction on the scaling factors in the NIfTI header.

Reviewer 3: I am linking this question to @HenkMutsaerts , who will probably be able to explain this much better than me!

Reviewer 5: @gllmflndn I also am not fond of this restriction. If I understood the earlier conversation, the reason that scale factors are complicated is that the DICOMs can have per-volume scale factors, and this is an issue with sequences where the M0, CBF and/or deltaM volumes are in the same series as the label/control volumes. To create a NIfTI with a shared slope/intercept, at least some of the volumes would have to be rescaled, introducing rounding error.

My personal preference would be to separate the label/control volumes from the M0/CBF/deltaM volumes during conversion, but I don't know how practicable that is.

(This is a rehash of https://github.com/bids-standard/bids-specification/pull/652#issuecomment-721488288, btw.)

Reviewer 3: @effigies Although I agree this might seems cleaner, I am afraid it is not practicle. 

The splitting up or not has been discussed many times with many ASL experts and all agreed on keeping the timeseries as acquired. The main disadvantage of ASL is that the way the data is stored (both the aslcontext within the timeseries, as scaling) is very much different between vendors, scanners, sequences, which is why flexibility is needed. 

I dug up the meeting notes on this discussion and these were the 2 main reasons why we chose for keeping the dataset in the original acquisition order:

- this allows reviewing temporal effects such as head motion (for which perfusion is very sensitive due to the low SNR)

- it allows flexibility which will be needed for more complex asl-approaches and labeling patterns. The current extension is focussing on these approaches which were included in the recommendations paper (Alsop 2015). However, more advanced approaches already exist and will probably gain more interest. Future updates of the ASL extensions might be needed and revisiting the basics of the ASL extension due to 'not flexible enough' would be inefficient.

I hope this helps in the discussion.

Reviewer 7: My personal preference would be the same than @effigies but if it has been thoroughly discussed and decided against, fine :smile: Despite this, I still don't see why there should be this statement in the specs: `and no scaling factors should be stored in the ancillary files or in the NIfTI header`. Once all of the data have been concatenated in a 4D file, in one way or another, the NIfTI scaling and data types could be anything and I don't see the need to be prescriptive here.

Again, apologies if I miss something obvious.

Reviewer 5: @patsycle Thanks for this explanation. That makes much more sense now.

I think I agree with @gllmflndn about this specific line, then. Whether to rescale or not feels like a decision that conversion software can handle without being dictated by BIDS.

Author: Sorry, for this understandable confusion. Appropriate analysis of ASL data is by design dependent on absolute quantification and having applied the correct scale factors (where this may not be an issue with other scan types). To avoid confusion if additional scale slopes - such as the MRScaleSlope in Philips, or the arbitrary factors that other vendors use - were or were not already applied, it seems most useful to put the correct scaling responsibility in the DICOM2BIDS conversion step. When ASL data is in the "BIDS standard" there should be no confusion of scale slopes left, and data should be fully ready for automatic processing.

Indeed, this doesn't mean that the NIfTI file cannot have a scaleslope nor that it has to be in double/floating-point per see, so we could remove this statement. However, that would require us to state something like: "ASL data should be properly scaled and any non-standard scale slopes - including those found in the private fields or not stored in the DICOMs - should be taken into account. Would you prefer this? I'm not sure if this would be confusing. Also, if we make floating point & NIfTI header scale slope "1" obliged, then it is easier to check if scale slopes were taken into account during conversion.

Reviewer 7: @HenkMutsaerts Thanks. I think it's just a matter of rephrasing. From the point of view of BIDS, there are no restrictions on the ASL NIfTI files. Then, you seem to suggest to add a note to DICOM2BIDS developers that data have to be scaled appropriately (potentially taking into account non-standard scale slopes found in DICOM private fields) to accommodate the unique scaling factor NIfTI files can have for all of the contained volumes.

Author: Yes, this makes sense! Thanks

The `m0scan` can either be stored inside the 4D ASL time-series NIfTI file or as a separate NIfTI file, depending on whether it 
was acquired within the ASL time-series or as a separate scan. These and other M0 options are specified in the required `M0` field of the `*_asl.json` file. 
It can also be stored under `fmap/sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.nii[.gz]`, 

> when the pepolar approach is used.

Reviewer 3: @Remi-Gau Link is added in the new PR!

> | PostLabelingDelay                 | REQUIRED                                                                                                                        | [number][] or [array][] of [numbers][]  | This is the postlabeling delay (PLD) time, in seconds, after the the end of the labeling (for `(P)CASL`) or middle of the labeling pulse (for `PASL`) until the middle of the excitation pulse applied to the imaging slab (for 3D acquisition) or first slice (for 2D acquisition).  Can be a number (for a single-PLD time series) or an array of numbers (for multi-PLD and Look-Locker). In the latter case, the array of numbers contains the PLD of each volume, namely each `control` and `label`, in the acquisition order. Any image within the time-series without a PLD, for example an `m0scan`, is indicated by a zero. Based on DICOM Tags 0018,9079 `Inversion Times` and 0018,0082 `InversionTime`.                                                                                                                                                                                                                                                                                                                                                                                                          |

Reviewer 7: There is a typo "after the THE end of the labeling..."

Reviewer 3: done!

> | EchoTime                          | REQUIRED                                                                                                                        | [number][] or [array][] of [numbers][]  | The echo time for the acquisition, specified in seconds. Specify either one number for the total time-series, or provide an array of numbers with different values of each volume. Corresponds to DICOM Tag 0018, 0081 `Echo Time` (please note that the DICOM term is in milliseconds not seconds).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer 3: @effigies Hi Chris, last friday we discussed the option to allow 'array of numbers', with the same length as number of volumes, for EchoTime and FlipAngles. I was wondering how we do this practically? I can easily propose it in the common fields in bep005, is that okay? For validator, should I ask Marco to do the same?

Reviewer 5: Yes, let's change it in **Timing Parameters** and **RF & Contrast**.

I would suggest the text:

> The echo time (TE) for the acquisition, specified in seconds. A single value applies to all volumes, or an array provides a value for each volume. Corresponds to DICOM Tag 0018, 0081 `Echo Time` (please note that the DICOM term is in milliseconds not seconds). 4D files with variable TE should only be used when volume timing is critical for interpretation of the data, as in ASL or variable TE fMRI sequences.

and

> Flip angle (FA) for the acquisition, specified in degrees. A single value applies to all volumes, or an array provides a value for each volume. Corresponds to: DICOM Tag 0018, 1314 `Flip Angle`. 4D files with variable FA should only be used when volume timing is critical for interpretation of the data, as in ASL or variable FA fMRI sequences.

Then in the ASL section, we mark them as REQUIRED for ASL sequences.

Reviewer 3: @effigies Thanks for the suggested texts. I will add those today. 

For Requirement level, I did change it now for MagneticFieldStrenght into 'RECOMENDED, but  REQUIRED for Arterial Spin Labeling', and the same for the new MRAcquisitionType field, in imitation of what was done already for EchoTime and SliceTiming. I can do the same for EchoTime and FlipAngle, but I agree that, if more BEP's do this, the requirement level fields might get long. What is prefered, so I can keep it consistent?

Reviewer 5: I think leaving "REQUIRED for Arterial Spin Labeling" is fine. We can reorganize in the future if things start getting hard to read.

Reviewer 3: I agree! I'll ask Marco to take a look into the validator as well for FlipAngle and EchoTime.

        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.json
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.json
```

The phase-encoding polarity (PEpolar) technique combines two or more EPI scans, with different phase encoding directions to estimate the underlying inhomogeneity/deformation 
map. In the case of arterial spin labeling, this usually concerns two ‘m0scans’. Examples of tools using this kind of images are FSL TOPUP, AFNI 3dqwarp and SPM. In such a case, 
the phase encoding direction is specified in the corresponding JSON file as one of: `i`, `j`, `k`, `i-`, `j-`,`k-`. For these differentially phase encoded sequences, one also 
needs to specify the Total Readout Time defined as the time (in seconds) from the center of the first echo to the center of the last echo (aka "FSL definition" - see
[here](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/topup/Faq#How_do_I_know_what_phase-encode_vectors_to_put_into_my_--datain_text_file.3F)

> and [here](https://lcni.uoregon.edu/kb-articles/kb-0003) how to calculate it). For example

Reviewer 5: I think this only added one sentence to the entire thing, but the reformatting makes it hard to be sure. Can you only modify lines that must be modified?

This will need to be updated with the latest changes in `master`.

Author: Thanks, there were several existing fields (e.g. FlipAngle, EchoTime, RepetitionTime, SliceTime) where we needed to add an option/case for ASL. The BIDS government committee advised us to copy these fields fully and adapt them, in the ASL section. We perhaps have to discuss these. For this case: @patsycle can you only change the parts that we adapted, and keep the original formatting, etc such that it is easier to read the difference?

Reviewer 3: @effigies indeed, we only added the ASL part, which is the sentence 'In the case of arterial .....two 'm0scans'. How should I change this practically?

Reviewer 5: Practically? Copy that sentence into your clipboard or somewhere convenient, merge `master` in and just accept `master`'s version of this chunk of text. Then, find where you'd put the sentence in the new version of the text as a separate commit.

Reviewer 3: @effigies Thanks. Just to be sure - this will probably merge all the other commits in other files from master to my branch - that is OK with you, right? If yes, then I will rebase master into my branch to keep my commits together.

    fmap/
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.json
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.json
```

The phase-encoding polarity (PEpolar) technique combines two or more EPI scans, with different phase encoding directions to estimate the underlying inhomogeneity/deformation 
map. In the case of arterial spin labeling, this usually concerns two ‘m0scans’. Examples of tools using this kind of images are FSL TOPUP, AFNI 3dqwarp and SPM. In such a case, 
the phase encoding direction is specified in the corresponding JSON file as one of: `i`, `j`, `k`, `i-`, `j-`,`k-`. For these differentially phase encoded sequences, one also 

> needs to specify the Total Readout Time defined as the time (in seconds) from the center of the first echo to the center of the last echo (aka "FSL definition" - see [here]
> (https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/topup/Faq#How_do_I_know_what_phase-encode_vectors_to_put_into_my_--datain_text_file.3F) and [here](https://lcni.uoregon.edu/kb-articles/kb-0003) how to calculate it). For example

Reviewer 2: ```suggestion
needs to specify the Total Readout Time defined as the time (in seconds) from the center of the first echo to the center of the last echo (aka "FSL definition" - see
[here](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/topup/Faq#How_do_I_know_what_phase-encode_vectors_to_put_into_my_--datain_text_file.3F)
and [here](https://lcni.uoregon.edu/kb-articles/kb-0003) how to calculate it). For example
```

    fmap/
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.json
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_m0scan.json
```

The phase-encoding polarity (PEpolar) technique combines two or more EPI scans, with different phase encoding directions to estimate the underlying inhomogeneity/deformation 
map. In the case of arterial spin labeling, this usually concerns two ‘m0scans’. Examples of tools using this kind of images are FSL TOPUP, AFNI 3dqwarp and SPM. In such a case, 
the phase encoding direction is specified in the corresponding JSON file as one of: `i`, `j`, `k`, `i-`, `j-`,`k-`. For these differentially phase encoded sequences, one also 
needs to specify the Total Readout Time defined as the time (in seconds) from the center of the first echo to the center of the last echo (aka "FSL definition" - see [here]

> (https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/topup/Faq#How_do_I_know_what_phase-encode_vectors_to_put_into_my_--datain_text_file.3F) and [here](https://lcni.uoregon.edu/kb-
> articles/kb-0003) how to calculate it). For example

Reviewer 2: you have a line break in the middle of the link --> this breaks the markdown syntax.

Note how there is a line break between `...egon.edu/kb-` and `articles/kb-0003) ...`

```suggestion
(https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/topup/Faq#How_do_I_know_what_phase-encode_vectors_to_put_into_my_--datain_text_file.3F) and [here](https://lcni.uoregon.edu/kb-articles/kb-0003) how to calculate it). For example
```

Reviewer 2: actually, also the `[here]` from the previous line needs to be on a single line with the following.

The markdown link syntax always needs to be on a single line: `[some text](url)` and NOT like this:

```
[this-link]
(http:www.wont-work.net)

[this-link](http:www.wont-work
-either.net)

[but-this-link](http:www.willwork.success)
```
